hydrolib/core/io/bc/models.py

```diff
@@ -236,6 +291,152 @@ class T3D(ForcingBase):
+            ):
+                raise ValueError(
+                    f"Vertical position index should be between 1 and {maximum_verticalpositionindex}"
+                )
+
+    @staticmethod
+    def _validate_verticalpositionindexes_and_update_quantityunitpairs(
+        verticalpositionindexes: List[int],
+        number_of_verticalpositions: int,
+        quantityunitpairs: List[QuantityUnitPair],
+    ) -> None:
+        if verticalpositionindexes is None:
+            raise ValueError("verticalpositionindex is not provided")
+
+        number_of_quantityunitpairs = len(quantityunitpairs)
+        number_of_verticalpositionindexes = len(verticalpositionindexes)
+
+        if number_of_verticalpositionindexes == number_of_quantityunitpairs:
+            T3D._validate_verticalpositionindexes_and_add_them_to_quantityunitpairs(
+                verticalpositionindexes[1:],
+                number_of_verticalpositions,
+                quantityunitpairs[1:],
+            )
+            return
+
+        if number_of_verticalpositionindexes == number_of_quantityunitpairs - 1:
+            T3D._validate_verticalpositionindexes_and_add_them_to_quantityunitpairs(
+                verticalpositionindexes,
+                number_of_verticalpositions,
+                quantityunitpairs[1:],
+            )
+            return
+
+        raise ValueError(
+            "Number of vertical positions should be equal to the number of units or equal to the number of units - 1"
+        )
+
+    @staticmethod
+    def _validate_verticalpositionindexes_and_add_them_to_quantityunitpairs(
+        verticalpositionindexes: List[int],
+        number_of_verticalpositions: int,
+        quantityunitpairs: List[QuantityUnitPair],
+    ) -> None:
+        T3D._validate_that_verticalpositionindexes_are_valid(
+            verticalpositionindexes, number_of_verticalpositions
+        )
+
+        T3D._add_verticalpositionindex_to_quantityunitpairs(
+            quantityunitpairs, verticalpositionindexes
+        )
+
+    @staticmethod
+    def _validate_that_verticalpositionindexes_are_valid(
+        verticalpositionindexes: List[int], number_of_vertical_positions: int
+    ) -> None:
+        for verticalpositionindexstring in verticalpositionindexes:
+            verticalpositionindex = (
+                int(verticalpositionindexstring)
+                if verticalpositionindexstring
+                else None
+            )
+            if not T3D._is_valid_verticalpositionindex(
+                verticalpositionindex, number_of_vertical_positions
+            ):
+                raise ValueError(
+                    f"Vertical position index should be between 1 and {number_of_vertical_positions}"
+                )
+
+    @staticmethod
+    def _is_valid_verticalpositionindex(
+        verticalpositionindex: int, number_of_vertical_positions: int
+    ) -> bool:
+        one_based_index_offset = 1
```

Comment:
not super useful? If (hypothetically) later it would be changed to `= 0` then still the expression below would need modification ` < number_of_vertical_positions`, instead of `<=`.
So, maybe directly put the `1` in the expression below?

Reply:
It is to clarify that we are dealing with a one-based index instead of the usual zero-based index.